--- a/drivers/net/ethernet/mediatek/Kconfig
+++ b/drivers/net/ethernet/mediatek/Kconfig
@@ -17,7 +17,6 @@ config NET_MEDIATEK_SOC
 	select PHYLINK
 	select DIMLIB
 	select PAGE_POOL
-	select PAGE_POOL_STATS
 	select PCS_MTK_LYNXI
 	select REGMAP_MMIO
 	help
--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -3595,6 +3595,7 @@
 
 static void mtk_ethtool_pp_stats(struct mtk_eth *eth, u64 *data)
 {
+#ifdef CONFIG_PAGE_POOL_STATS
 	struct page_pool_stats stats = {};
 	int i;
 
@@ -3607,6 +3607,7 @@
 		page_pool_get_stats(ring->page_pool, &stats);
 	}
 	page_pool_ethtool_stats_get(data, &stats);
+#endif
 }
 
 static void mtk_get_ethtool_stats(struct net_device *dev,
