--- a/include/linux/netdevice.h
+++ b/include/linux/netdevice.h
@@ -882,8 +882,8 @@ struct net_device_path {
 	};
 };
 
-#define NET_DEVICE_PATH_STACK_MAX	5
-#define NET_DEVICE_PATH_VLAN_MAX	2
+#define NET_DEVICE_PATH_STACK_MAX	8
+#define NET_DEVICE_PATH_VLAN_MAX	3
 
 struct net_device_path_stack {
 	int			num_paths;
--- a/include/net/netfilter/nf_flow_table.h
+++ b/include/net/netfilter/nf_flow_table.h
@@ -96,7 +96,7 @@ enum flow_offload_xmit_type {
 	FLOW_OFFLOAD_XMIT_DIRECT,
 };
 
-#define NF_FLOW_TABLE_ENCAP_MAX		2
+#define NF_FLOW_TABLE_ENCAP_MAX		3
 
 struct flow_offload_tuple {
 	union {
