--- a/netfilter/nf_flow_table_ip.c
+++ b/netfilter/nf_flow_table_ip.c
@@ -158,7 +158,7 @@ static void nf_flow_tuple_encap(struct s
 		tuple->encap[i].proto = skb->protocol;
 		break;
 	case htons(ETH_P_PPP_SES):
-		phdr = (struct pppoe_hdr *)skb_mac_header(skb);
+		phdr = (struct pppoe_hdr *)(skb_mac_header(skb) + ETH_HLEN);
 		tuple->encap[i].id = ntohs(phdr->sid);
 		tuple->encap[i].proto = skb->protocol;
 		break;
