From 1de6bbc10872b67f9f54a2511e6750a6071f5544 Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Fri, 19 Mar 2021 22:24:08 +0000
Subject: [PATCH 1/6] Fix FTBS on FreeBSD due to Linux-specific optimisation of
 if_nametoindex()

---
 src/network.c | 16 ++++++++++------
 1 file changed, 10 insertions(+), 6 deletions(-)

--- a/src/network.c
+++ b/src/network.c
@@ -720,13 +720,17 @@ int enumerate_interfaces(int reset)
      to a server is specified by an interface, so cache them.
      Update the cache here. */
   for (serv = daemon->servers; serv; serv = serv->next)
-    if (strlen(serv->interface) != 0)
+    if (serv->interface[0] != 0)
       {
-	 struct ifreq ifr;
-
-	 safe_strncpy(ifr.ifr_name, serv->interface, IF_NAMESIZE);
-	 if (ioctl(param.fd, SIOCGIFINDEX, &ifr) != -1) 
-	   serv->ifindex = ifr.ifr_ifindex;
+#ifdef HAVE_LINUX_NETWORK
+	struct ifreq ifr;
+	
+	safe_strncpy(ifr.ifr_name, serv->interface, IF_NAMESIZE);
+	if (ioctl(param.fd, SIOCGIFINDEX, &ifr) != -1) 
+	  serv->ifindex = ifr.ifr_ifindex;
+#else
+	serv->ifindex = if_nametoindex(serv->interface);
+#endif
       }
     
 again:
