From b082842ee7af9f91a4752024ac2d9ea4da0e878a Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Mon, 12 Apr 2021 17:43:59 +0100
Subject: [PATCH 05/10] Add NO_LOOP to config.h docs.

---
 src/config.h | 1 +
 1 file changed, 1 insertion(+)

--- a/src/config.h
+++ b/src/config.h
@@ -143,6 +143,7 @@ NO_SCRIPT
 NO_LARGEFILE
 NO_AUTH
 NO_DUMPFILE
+NO_LOOP
 NO_INOTIFY
    these are available to explicitly disable compile time options which would 
    otherwise be enabled automatically or which are enabled  by default 
