From ffa4628faa5af2f8cf11e343bf0dfde1b88ba488 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?=E9=BB=8E=E9=86=92=E8=81=AA?= <lixingcong512@gmail.com>
Date: Mon, 22 Mar 2021 22:00:26 +0000
Subject: [PATCH 6/6] Fix thinko in 51f7bc924cbcdeb09cbb83249b70c121d1ffa31e

---
 src/forward.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/src/forward.c
+++ b/src/forward.c
@@ -2003,7 +2003,7 @@ unsigned char *tcp_request(int confd, ti
 			    }
 			  
 #ifdef MSG_FASTOPEN
-			    server_send(last_server, last_server->tcpfd, packet, m + sizeof(u16), MSG_FASTOPEN);
+			    server_send(last_server, last_server->tcpfd, packet, size + sizeof(u16), MSG_FASTOPEN);
 
 			    if (errno == 0)
 			      data_sent = 1;
