From c2f129ba3dbc21828adaa296f3edb9edafa9e29f Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Tue, 11 Jan 2022 22:48:14 +0000
Subject: [PATCH 09/16] Fix FTBFS when HAVE_DNSSEC not defined.

---
 src/forward.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/src/forward.c b/src/forward.c
index 1ed8cff..8d8b8fe 100644
--- a/src/forward.c
+++ b/src/forward.c
@@ -19,7 +19,9 @@
 static struct frec *get_new_frec(time_t now, struct server *serv, int force);
 static struct frec *lookup_frec(unsigned short id, int fd, void *hash, int *firstp, int *lastp);
 static struct frec *lookup_frec_by_query(void *hash, unsigned int flags, unsigned int flagmask);
+#ifdef HAVE_DNSSEC
 static struct frec *lookup_frec_dnssec(char *target, int class, int flags, struct dns_header *header);
+#endif
 
 static unsigned short get_id(void);
 static void free_frec(struct frec *f);
@@ -2629,6 +2631,7 @@ static struct frec *lookup_frec_by_query(void *hash, unsigned int flags, unsigne
   return NULL;
 }
 
+#ifdef HAVE_DNSSEC
 /* DNSSEC frecs have the complete query in the block stash.
    Search for an existing query using that. */
 static struct frec *lookup_frec_dnssec(char *target, int class, int flags, struct dns_header *header)
@@ -2657,7 +2660,7 @@ static struct frec *lookup_frec_dnssec(char *target, int class, int flags, struc
 
    return NULL;
 }
-
+#endif
 
 /* Send query packet again, if we can. */
 void resend_query()
-- 
2.34.1

