From d100eb05a30c2891c00afa12af7753f60518ffa4 Mon Sep 17 00:00:00 2001
From: Simon Kelley <simon@thekelleys.org.uk>
Date: Wed, 14 Apr 2021 21:15:34 +0100
Subject: [PATCH 10/10] <nettle/version.h> -> <nettle/bignum.h> for backward
 compatibility.

---
 src/crypto.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

--- a/src/crypto.c
+++ b/src/crypto.c
@@ -19,7 +19,10 @@
 #if defined(HAVE_DNSSEC) || defined(HAVE_CRYPTOHASH)
 
 /* Minimal version of nettle */
-#include <nettle/version.h>
+
+/* bignum.h includes version.h and works on
+   earlier releases of nettle which don't have version.h */
+#include <nettle/bignum.h>
 #if !defined(NETTLE_VERSION_MAJOR)
 #  define NETTLE_VERSION_MAJOR 2
 #  define NETTLE_VERSION_MINOR 0
