--- a/Makefile
+++ b/Makefile
@@ -723,3 +723,5 @@ KBUILD_CFLAGS	+= $(call cc-disable-warni
 else ifdef CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE_O3
 KBUILD_CFLAGS += -O3 $(EXTRA_OPTIMIZATION)
+else ifdef CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE_OFAST
+KBUILD_CFLAGS += -Ofast $(EXTRA_OPTIMIZATION)
 else ifdef CONFIG_CC_OPTIMIZE_FOR_SIZE
--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1210,12 +1210,18 @@
 config CC_OPTIMIZE_FOR_PERFORMANCE_O3
 	bool "Optimize more for performance (-O3)"
 	depends on ARC
 	help
 	  Choosing this option will pass "-O3" to your compiler to optimize
 	  the kernel yet more for performance.
 
+config CC_OPTIMIZE_FOR_PERFORMANCE_OFAST
+	bool "Optimize more for maximum performance (-Ofast)"
+	help
+	  Choosing this option will pass "-Ofast" to your compiler to optimize
+	  the kernel yet more for performance.
+
 config CC_OPTIMIZE_FOR_SIZE
 	bool "Optimize for size (-Os)"
 	help
 	  Choosing this option will pass "-Os" to your compiler resulting
 	  in a smaller kernel.

