--- a/dev.c
+++ b/dev.c
@@ -206,6 +206,7 @@ void qmi_request_cancel(struct qmi_dev *
 int qmi_request_wait(struct qmi_dev *qmi, struct qmi_request *req)
 {
 	bool complete = false;
+	bool *c;
 	bool cancelled;
 
 	if (!req->pending)
@@ -226,8 +227,10 @@ int qmi_request_wait(struct qmi_dev *qmi
 		uloop_cancelled = cancelled;
 	}
 
-	if (req->complete == &complete)
-		req->complete = NULL;
+	c = req->complete;
+	req->complete = NULL;
+	if (c != &complete)
+		req->complete = c;
 
 	return req->ret;
 }
