--- a/libopkg/opkg_conf.h
+++ b/libopkg/opkg_conf.h
@@ -80,6 +80,8 @@
 	int check_signature;
 	int force_signature;
 	int no_check_certificate;
+	int restrict_ip_proto;
+	char *not_upgradable_mask;
 	int nodeps;		/* do not follow dependencies */
 	int nocase;		/* perform case insensitive matching */
 	char *offline_root;
--- a/libopkg/opkg_conf.c
+++ b/libopkg/opkg_conf.c
@@ -55,6 +55,8 @@
 	{"force_checksum", OPKG_OPT_TYPE_BOOL, &_conf.force_checksum},
 	{"check_signature", OPKG_OPT_TYPE_BOOL, &_conf.check_signature},
 	{"no_check_certificate", OPKG_OPT_TYPE_BOOL, &_conf.no_check_certificate},
+	{"restrict_ip_proto", OPKG_OPT_TYPE_INT, &_conf.restrict_ip_proto},
+	{"not_upgradable_mask", OPKG_OPT_TYPE_STRING, &_conf.not_upgradable_mask},
 	{"ftp_proxy", OPKG_OPT_TYPE_STRING, &_conf.ftp_proxy},
 	{"http_proxy", OPKG_OPT_TYPE_STRING, &_conf.http_proxy},
 	{"http_timeout", OPKG_OPT_TYPE_STRING, &_conf.http_timeout},
--- a/libopkg/opkg_download.c
+++ b/libopkg/opkg_download.c
@@ -156,6 +156,12 @@
 		if (conf->no_check_certificate) {
 			argv[i++] = "--no-check-certificate";
 		}
+		if (conf->restrict_ip_proto == 4) {
+			argv[i++] = "-4";
+		}
+		if (conf->restrict_ip_proto == 6) {
+			argv[i++] = "-6";
+		}
 		if (conf->http_timeout) {
 			argv[i++] = "--timeout";
 			argv[i++] = conf->http_timeout;
