From a69773e1c2c55a8c8e741616830aa137930ff4ab Mon Sep 17 00:00:00 2001
From: rodrigo <rodrigo.sousa.577@gmail.com>
Date: Mon, 31 Oct 2022 22:46:21 +0000
Subject: [PATCH] Atualizar 'mt7615/eeprom.h'

---
 mt7615/mt7615.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/mt7615/mt7615.h b/mt7615/mt7615.h
index 060d52c8..eff77dae 100644
--- a/mt7615/mt7615.h
+++ b/mt7615/mt7615.h
@@ -50,7 +50,7 @@
 #define MT7663_ROM_PATCH		"mediatek/mt7663pr2h_rebb.bin"
 #define MT7663_FIRMWARE_N9		"mediatek/mt7663_n9_rebb.bin"
 
-#define MT7615_EEPROM_SIZE		1024
+#define MT7615_EEPROM_SIZE		1536
 #define MT7663_EEPROM_SIZE		1536
 #define MT7615_TOKEN_SIZE		4096
 
