--- a/drivers/char/lrng/lrng_es_mgr.c
+++ b/drivers/char/lrng/lrng_es_mgr.c
@@ -7,7 +7,6 @@
 #include <linux/random.h>
 #include <linux/utsname.h>
 #include <linux/workqueue.h>
-#include <asm/archrandom.h>

 #include "lrng_drng_atomic.h"
 #include "lrng_drng_mgr.h"
--- a/drivers/char/lrng/lrng_es_cpu.c
+++ b/drivers/char/lrng/lrng_es_cpu.c
@@ -7,7 +7,6 @@
 #include <crypto/hash.h>
 #include <linux/module.h>
 #include <linux/random.h>
-#include <asm/archrandom.h>

 #include "lrng_definitions.h"
 #include "lrng_es_aux.h"
