--- a/lib/zstd/compress.c
+++ b/lib/zstd/compress.c
@@ -3290,5 +3290,5 @@
 /*-=====  Pre-defined compression levels  =====-*/
 
-#define ZSTD_DEFAULT_CLEVEL 1
-#define ZSTD_MAX_CLEVEL 22
+#define ZSTD_DEFAULT_CLEVEL 23
+#define ZSTD_MAX_CLEVEL 23
 int ZSTD_maxCLevel(void) { return ZSTD_MAX_CLEVEL; }
@@ -3322,2 +3322,3 @@
     {27, 27, 25, 9, 3, 512, ZSTD_btopt2}, /* level 22 */
+    {20, 16, 17, 21, 2, 1024, ZSTD_btopt2}, /* level 23 */
     },
@@ -3349,2 +3350,3 @@
     {18, 19, 18, 13, 3, 512, ZSTD_btopt2}, /* level 22.*/
+    {18, 16, 15, 21, 2, 1024, ZSTD_btopt2}, /* level 23.*/
     },
@@ -3376,2 +3378,3 @@
     {17, 18, 17, 11, 3, 512, ZSTD_btopt2}, /* level 22.*/
+    {17, 12, 12, 21, 2, 1024, ZSTD_btopt2}, /* level 23.*/
     },
@@ -3403,2 +3406,3 @@
     {14, 15, 15, 10, 3, 256, ZSTD_btopt2}, /* level 22.*/
+    {14, 12, 12, 21, 2, 1024, ZSTD_btopt2}, /* level 23.*/
     },
--- a/crypto/zstd.c
+++ b/crypto/zstd.c
@@ -23,6 +23,6 @@
 #include <crypto/internal/scompress.h>
 
 
-#define ZSTD_DEF_LEVEL	3
+#define ZSTD_DEF_LEVEL	23
 
 struct zstd_ctx {
