--- a/init/Kconfig
+++ b/init/Kconfig
@@ -1365,10 +1365,15 @@
 
 config CC_OPTIMIZE_FOR_PERFORMANCE_O3
 	bool "Optimize more for performance (-O3)"
-	depends on ARC
 	help
 	  Choosing this option will pass "-O3" to your compiler to optimize
 	  the kernel yet more for performance.
+
+config CC_OPTIMIZE_FOR_PERFORMANCE_OFAST
+	bool "Optimize for maximum performance (-Ofast)"
+	help
+	  Choosing this option will pass "-Ofast" to your compiler to optimize
+	  the kernel for maximum performance.
 
 config CC_OPTIMIZE_FOR_SIZE
 	bool "Optimize for size (-Os)"
--- a/Makefile
+++ b/Makefile
@@ -755,6 +755,8 @@
 KBUILD_CFLAGS += -O2 $(EXTRA_OPTIMIZATION)
 else ifdef CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE_O3
 KBUILD_CFLAGS += -O3 $(EXTRA_OPTIMIZATION)
+else ifdef CONFIG_CC_OPTIMIZE_FOR_PERFORMANCE_OFAST
+KBUILD_CFLAGS += -Ofast $(EXTRA_OPTIMIZATION)
 else ifdef CONFIG_CC_OPTIMIZE_FOR_SIZE
 KBUILD_CFLAGS += -Os -fno-reorder-blocks -fno-tree-ch $(EXTRA_OPTIMIZATION)
 endif
